File: atividade-15/atividade.go
package main

import "fmt"

type AFN struct {
	startState   string
	acceptStates map[string]bool
	currentState string
}

func NewAFN() *AFN {
	return &AFN{
		startState:   "q0",
		acceptStates: map[string]bool{"q0": true},
		currentState: "q0",
	}
}

func (afn *AFN) transition() {
	if afn.currentState == "q0" {
		afn.currentState = "q1"
	} else if afn.currentState == "q1" {
		afn.currentState = "q0"
	}
}

func (afn *AFN) isAccepted(input string) bool {
	afn.currentState = afn.startState

	for range input {
		afn.transition()
	}

	
	return afn.acceptStates[afn.currentState]
}

func main() {
	afn := NewAFN()
	fmt.Println(afn.isAccepted("ab"))    
	fmt.Println(afn.isAccepted("abab"))  
	fmt.Println(afn.isAccepted("a"))     
	fmt.Println(afn.isAccepted("abc"))   
}

File: atividade-15/atividade.py
class AFN:
    def __init__(self):
        self.states = {'q0', 'q1'}  
        self.start_state = 'q0'  
        self.accept_states = {'q0'}  

    def is_accepted(self, string):
        current_states = {self.start_state}  

        for char in string:
            next_states = set()
            for state in current_states:
                if state == 'q0':
                    next_states.add('q1')  
                elif state == 'q1':
                    next_states.add('q0')  
            current_states = next_states

       
        return 'q0' in current_states

if __name__ == '__main__':
    afn = AFN()
    print(afn.is_accepted("ab"))    
    print(afn.is_accepted("abab"))  
    print(afn.is_accepted("a"))     
    print(afn.is_accepted("abc"))   
